### PR TITLE
Variables

### PR DESCRIPTION
**Pull request by hooksie1**
Started with just variables and it turned into a little more.

### CODE REVIEW

**Comment by hooksie1** [comment]
Depending how much devel you are going to do with this you might want to put it on Gitlab. The automated CI is really nice.

**Comment by Danp2** [comment]
Does adding the full path to the binaries make the script less compatible with various distros?

**Comment by hooksie1** [comment]
It should be the same between upstream/downstream distros. I think the benefits of security and assurance outweigh that issue. Especially since there is a cd into /opt. If someone has /opt in their PATH and any custom commands it could call those first instead of the intended utility.

**Comment by Danp2** [comment]
What about Debian/Centos? Are the binary files stored in the same location?

**Comment by hooksie1** [comment]
That's what I meant by upstream/downstream. Debian and Ubuntu should be the same. I didn't think XO was supported on CentOS at all?

**Comment by Jarli01** [comment]
@hooksie1 technically CentOS isn't excluded from being used, but the developers work on Ubuntu, so why divert from that when the goal is to mimic (as much as possible) the XOA version. It limits the need for special support, and anyone from us to the dev team can easily assume that everything is Ubuntu based.

**Comment by Jarli01** [comment]
I'm testing the process now on a new installation, will update / merge shortly.

**Comment by Jarli01** [comment]
Something with the changes you made to the xo-server.service fails, investigating.

**Comment by hooksie1** [comment]
That was my point about CentOS.  Bah I fixed the mistake. I had one variable set up for it and split it into two (one for the dir and one for the file) and forgot to change it.

File: xo_install.sh [hunk 1]
```diff
@@ -1,24 +1,46 @@
...
 sudo cp sample.config.yaml .xo-server.yaml
-sudo sed -i /mounts/a\\"    '/': '/opt/xo-web/dist'" .xo-server.yaml
-cd /opt/xo-web
-yarn install --force
-cat > /etc/systemd/system/xo-server.service <<EOF
+sudo sed -i "s|#'/': '/path/to/xo-web/dist/'|'/': '/opt/xo-web/dist'|" .xo-server.yaml
+cd $xo_web_dir
+/usr/bin/yarn install --force
+
+if [[ ! -e $xo_systemd_service ]] ; then
```
Comment (Danp2) [inline xo_install.sh]: Need to update other instance as well. ;-)

File: xo_install.sh [hunk 2]
```diff
@@ -1,24 +1,46 @@
...
+xo_server="https://github.com/vatesfr/xo-server"
+xo_web="https://github.com/vatesfr/xo-web"
+n_repo="https://raw.githubusercontent.com/visionmedia/n/master/bin/n"
+yarn_repo="deb https://dl.yarnpkg.com/debian/ stable main"
+node_source="https://deb.nodesource.com/setup_5.x"
+yarn_gpg="https://dl.yarnpkg.com/debian/pubkey.gpg"
+n_location="/usr/local/bin/n"
+xo_server_dir="/opt/xo-server"
+xo_web_dir="/opt/xo-web"
+systemd_service_dir="/lib/systemd/system"
```
Comment (hooksie1) [inline xo_install.sh]: /etc/systemd/system is usually just symlinks and .wants. Default is /lib/systemd/system.

File: xo_install.sh (hunk 1 above)
Comment (hooksie1) [inline xo_install.sh]: yup my bad.

Comment (Jarli01) [inline xo_install.sh]: @hooksie1 we're calling xo-server.service you're naming it xo_service and it fails to execute. 

once the build is complete.

Comment (hooksie1) [inline xo_install.sh]: Made a new commit that fixed this when Dan pointed it out.

**Comment by Jarli01** [comment]
`Aug 03 14:21:29 xo systemd[1]: Started XO Server.
Aug 03 14:21:29 xo xo-server[9322]: module.js:487
Aug 03 14:21:29 xo xo-server[9322]:     throw err;
Aug 03 14:21:29 xo xo-server[9322]:     ^
Aug 03 14:21:29 xo xo-server[9322]: Error: Cannot find module 'bluebird'
Aug 03 14:21:29 xo xo-server[9322]:     at Function.Module._resolveFilename (module.js:485:15)
Aug 03 14:21:29 xo xo-server[9322]:     at Function.Module._load (module.js:437:25)
Aug 03 14:21:29 xo xo-server[9322]:     at Module.require (module.js:513:17)
Aug 03 14:21:29 xo xo-server[9322]:     at require (internal/module.js:11:18)
Aug 03 14:21:29 xo xo-server[9322]:     at Object.<anonymous> (/opt/xo-server/bin/xo-server:12:18)
Aug 03 14:21:29 xo xo-server[9322]:     at Module._compile (module.js:569:30)
Aug 03 14:21:29 xo xo-server[9322]:     at Object.Module._extensions..js (module.js:580:10)
Aug 03 14:21:29 xo xo-server[9322]:     at Module.load (module.js:503:32)
Aug 03 14:21:29 xo xo-server[9322]:     at tryModuleLoad (module.js:466:12)
Aug 03 14:21:29 xo xo-server[9322]:     at Function.Module._load (module.js:458:3)
Aug 03 14:21:29 xo xo-server[9322]:     at Function.Module.runMain (module.js:605:10)
Aug 03 14:21:29 xo xo-server[9322]:     at startup (bootstrap_node.js:158:16)
Aug 03 14:21:29 xo xo-server[9322]:     at bootstrap_node.js:575:3
Aug 03 14:21:29 xo systemd[1]: xo-server.service: Main process exited, code=exited, status=1/FAILURE
Aug 03 14:21:29 xo systemd[1]: xo-server.service: Unit entered failed state.
Aug 03 14:21:29 xo systemd[1]: xo-server.service: Failed with result 'exit-code'.
Aug 03 14:21:29 xo systemd[1]: xo-server.service: Service hold-off time over, scheduling restart.
Aug 03 14:21:29 xo systemd[1]: Stopped XO Server.
Aug 03 14:21:29 xo systemd[1]: Started XO Server.
Aug 03 14:21:30 xo xo-server[9349]: module.js:487
Aug 03 14:21:30 xo xo-server[9349]:     throw err;
Aug 03 14:21:30 xo xo-server[9349]:     ^
Aug 03 14:21:30 xo xo-server[9349]: Error: Cannot find module 'bluebird'
Aug 03 14:21:30 xo xo-server[9349]:     at Function.Module._resolveFilename (module.js:485:15)
Aug 03 14:21:30 xo xo-server[9349]:     at Function.Module._load (module.js:437:25)
Aug 03 14:21:30 xo xo-server[9349]:     at Module.require (module.js:513:17)
Aug 03 14:21:30 xo xo-server[9349]:     at require (internal/module.js:11:18)
Aug 03 14:21:30 xo xo-server[9349]:     at Object.<anonymous> (/opt/xo-server/bin/xo-server:12:18)
Aug 03 14:21:30 xo xo-server[9349]:     at Module._compile (module.js:569:30)
Aug 03 14:21:30 xo xo-server[9349]:     at Object.Module._extensions..js (module.js:580:10)
Aug 03 14:21:30 xo xo-server[9349]:     at Module.load (module.js:503:32)
Aug 03 14:21:30 xo xo-server[9349]:     at tryModuleLoad (module.js:466:12)
Aug 03 14:21:30 xo xo-server[9349]:     at Function.Module._load (module.js:458:3)
Aug 03 14:21:30 xo xo-server[9349]:     at Function.Module.runMain (module.js:605:10)
Aug 03 14:21:30 xo xo-server[9349]:     at startup (bootstrap_node.js:158:16)
Aug 03 14:21:30 xo xo-server[9349]:     at bootstrap_node.js:575:3
Aug 03 14:21:30 xo systemd[1]: xo-server.service: Main process exited, code=exited, status=1/FAILURE
Aug 03 14:21:30 xo systemd[1]: xo-server.service: Unit entered failed state.
Aug 03 14:21:30 xo systemd[1]: xo-server.service: Failed with result 'exit-code'.
Aug 03 14:21:30 xo systemd[1]: xo-server.service: Service hold-off time over, scheduling restart.
Aug 03 14:21:30 xo systemd[1]: Stopped XO Server.
Aug 03 14:21:30 xo systemd[1]: xo-server.service: Start request repeated too quickly.
Aug 03 14:21:30 xo systemd[1]: Failed to start XO Server.
Aug 03 14:21:30 xo systemd[1]: xo-server.service: Unit entered failed state.
Aug 03 14:21:30 xo systemd[1]: xo-server.service: Failed with result 'exit-code'.
`

Is what I'm getting on a clean install.

**Comment by Jarli01** [comment]
Looks like we might [need to keep](https://stackoverflow.com/questions/31378699/node-js-error-error-cannot-find-module-bluebird) the node.js version 'bluebird'

**Comment by Danp2** [comment]
Seeing the same error. Tried building using master script without recent changes and it worked correctly.

**Comment by Jarli01** [comment]
Same, using the original no issues with the install. 

Yet I'm not seeing the specific node.js version which this installer is complaining about.

**Comment by Danp2** [comment]
The entire node_modules directory is missing from /opt/xo-server

**Comment by Danp2** [comment]
Running `yarn install --force` instead of `sudo /usr/bin/npm install && /usr/bin/npm run build` seems to fix the issue.

**Comment by Jarli01** [comment]
I looked in /opt/xo-server and didn't even notice that. 

Good find.

**Comment by Danp2** [comment]
I created a separate PR to address the issue. Please review when you have a moment.

**Comment by Danp2** [comment]
@Jarli01 Builds correctly now. Need to resolve conflicts, which only you have authority to do. 😄

**Pull request merged**